patch.go

```diff
@@ -0,0 +1,465 @@
+	var ops []patchOperation
+	for i, e := range arr.Elements {
+		obj, ok := e.Value.(*Object)
+		if !ok {
+			return nil, fmt.Errorf("hujson: patch operation %d: must be a JSON object", i)
+		}
+		seen := make(map[string]bool)
+		var op patchOperation
+		for j, m := range obj.Members {
+			name := m.Name.Value.(Literal).String()
+			if seen[name] {
+				return nil, fmt.Errorf("hujson: patch operation %d: duplicate name %q", i, m.Name.Value)
+			}
+			seen[name] = true
+			switch name {
+			case "op":
+				if m.Value.Value.Kind() != '"' {
+					return nil, fmt.Errorf("hujson: patch operation %d: member %q must be a JSON string", i, name)
+				}
+				switch opType := m.Value.Value.(Literal).String(); opType {
+				case "add", "remove", "replace", "move", "copy", "test":
+					op.op = opType
+				default:
+					return nil, fmt.Errorf("hujson: patch operation %d: unknown operation %q", i, m.Value.Value)
+				}
+			case "path":
+				if m.Value.Value.Kind() != '"' {
+					return nil, fmt.Errorf("hujson: patch operation %d: member %q must be a JSON string", i, name)
+				}
+				op.path = m.Value.Value.(Literal).String()
+			case "from":
+				if m.Value.Value.Kind() != '"' {
+					return nil, fmt.Errorf("hujson: patch operation %d: member %q must be a JSON string", i, name)
+				}
+				op.from = m.Value.Value.(Literal).String()
+			case "value":
+				m.Value.BeforeExtra = obj.beforeExtraAt(j + 0).extractLeadingComments(true)
+				m.Value.AfterExtra = obj.beforeExtraAt(j + 1).extractTrailingcomments(true)
+				op.value = m.Value
+			}
+		}
+		switch {
+		case !seen["op"]:
+			return nil, fmt.Errorf("hujson: patch operation %d: missing required member %q", i, "op")
+		case !seen["path"]:
+			return nil, fmt.Errorf("hujson: patch operation %d: missing required member %q", i, "path")
+		case !seen["from"] && (op.op == "move" || op.op == "copy"):
+			return nil, fmt.Errorf("hujson: patch operation %d: missing required member %q", i, "from")
+		case !seen["value"] && (op.op == "add" || op.op == "replace" || op.op == "test"):
+			return nil, fmt.Errorf("hujson: patch operation %d: missing required member %q", i, "value")
+		}
+		ops = append(ops, op)
+	}
+	return ops, nil
+}
+
+func (v *Value) patchAdd(i int, op patchOperation) error {
+	switch s, err := v.find(findState{pointer: op.path}); {
+	case err == nil || (err == errNotFound && len(s.pointer) == s.offset):
+		if s.parent == nil {
+			*v = op.value // only occurs for root
+		} else {
+			switch comp := s.parent.(type) {
+			case *Object:
+				if s.idx < comp.length() {
+					replaceAt(comp, s.idx, op.value)
+				} else {
+					insertAt(comp, s.idx, op.value)
+					comp.Members[s.idx].Name.Value = String(s.name)
+				}
+			case *Array:
+				insertAt(comp, s.idx, op.value)
+			}
+		}
+		return nil
+	default:
+		return fmt.Errorf("hujson: patch operation %d: %v", i, err)
+	}
+}
+
+func (v *Value) patchRemoveOrReplace(i int, op patchOperation) error {
+	switch s, err := v.find(findState{pointer: op.path}); {
```

Comment:
I'm curious what led you to this style. It seems a little unusual. Go would typically read:

```
s, err := v.find(...)
if err != nil {
    return fmt.Errorf("...")
}
/// the err == nil case
```

Reply:
Now that I look at it, I'm not sure. At least I was consistent about my excessive usage of `switch` statements :)

I'll make it more conventional.